chat-plugins/poll.js

```diff
@@ -0,0 +1,155 @@
+* users that have voted, and are updated in real time. The poll can be closed with /endpoll.
+* By bumbadadabum with (a lot of) help from Zarel.
+*/
+
+var Poll = (function () {
+    function Poll(room, question, options) {
+        this.room = room;
+        this.question = question;
+        this.voters = [];
+        this.totalVotes = 0;
+
+        this.options = [];
+        for (var i in options) {
+            this.options.push({name: options[i], votes: 0});
+        }
+    }
+
+    Poll.prototype.vote = function (user, option) {
+        for (var ip in user.ips) {
+            if (this.voters.indexOf(ip) > -1) {
+                return this.room.send('You have already voted for this poll.');
+            }
+            else {
+                this.voters.push(ip);
+            }
+        }
+
+        this.options[option - 1].votes++;
+        this.totalVotes++;
+
+        this.update();
+    };
+
+    Poll.prototype.generateVotes = function () {
+        var output = '<div class="infobox" style="text-align:center"><p style="font-weight:bold;font-size:14pt">' + this.question + '</p>';
+        for (var i in this.options) {
+            output += '<button value="/votepoll ' + (parseInt(i) + 1) + '" name="send">' + (parseInt(i) + 1) + '. ' + this.options[i].name + '</button><br/>';
+        }
+        output += '</div>';
+
+        return output;
+    };
+
+    Poll.prototype.generateResults = function () {
+        var output = '<div class="infobox" style="text-align:center"><p style="font-weight:bold;font-size:14pt">' + this.question + '</p>';
+        for (var i in this.options) {
+            output += '<span style="font-weight:bold;padding:2px">' + (parseInt(i) + 1) + '. ' + this.options[i].name + '</span><br/><span style="font-size:8pt;padding:0px">' + this.options[i].votes + ' votes (' + ((this.options[i].votes * 100) / this.totalVotes) + '% of total).</span><br/>';
+        }
+        output += '</div>';
+
+        return output;
+    };
+
+    Poll.prototype.update = function () {
+        var results = this.generateResults();
+
+        // Update the poll results for everyone that has voted
+        for (var i in this.room.users) {
+            var user = this.room.users[i];
+            if (this.voters.indexOf(user.latestIp) >= 0) {
+                user.sendTo(this.room, '|uhtmlchange|poll|' + results);
+            }
+        }
+    };
+
+    Poll.prototype.display = function (user, broadcast) {
+        var votes = this.generateVotes();
+        var results = this.generateResults();
+
+        var target = [];
+
+        if (broadcast) {
+            target = this.room.users;
+        } else {
+            target.push(user);
+        }
+
+        for (var i in target) {
+            var thisUser = target[i];
+            if (this.voters.indexOf(thisUser.latestIp) >= 0) {
+                thisUser.sendTo(this.room, '|uhtml|poll|' + results);
+            }
+            else {
```

Comment:
``else`` should be in the line above...

TODO: Fix conditional validation in build...